crates/nu-cli/src/commands/cal.rs

```diff
@@ -66,47 +66,129 @@ pub fn cal(args: CommandArgs, registry: &CommandRegistry) -> Result<OutputStream
-            if let Ok(year_u64) = year.as_u64() {
-                year_value = year_u64;
-            }
+        if let Some(year_value) = args.get("full-year") {
+            if let Ok(year_u64) = year_value.as_u64() {
+                selected_year = year_u64 as i32;
 
-            if year_value != current_year as u64 {
-                day_value = None
+                if selected_year != current_year {
+                    current_day_option = None
+                }
             }
         }
 
+        let (month_helper, _) = MonthHelper::new(selected_year, current_year, current_month);
+
         add_year_to_table(
+            &args,
             &mut calendar_vec_deque,
             &tag,
-            year_value as i32,
-            current_year,
-            current_month,
-            day_value,
-            &args,
+            &month_helper,
+            current_day_option,
         );
     } else {
-        let (day_start_offset, number_of_days_in_month, _) =
-            get_month_information(current_year, current_month, current_year);
+        let (month_helper, _) = MonthHelper::new(current_year, current_year, current_month);
 
         add_month_to_table(
+            &args,
             &mut calendar_vec_deque,
             &tag,
-            current_year,
-            current_month,
+            &month_helper,
             Some(current_day),
-            day_start_offset,
-            number_of_days_in_month as usize,
-            &args,
         );
     }
 
     Ok(futures::stream::iter(calendar_vec_deque).to_output_stream())
 }
 
+struct MonthHelper {
+    day_number_month_starts_on: usize,
+    number_of_days_in_month: usize,
+    selected_year: i32,
+    default_year_if_error: i32,
+    month: u32,
+}
+
+impl MonthHelper {
+    pub fn new(selected_year: i32, default_year_if_error: i32, month: u32) -> (MonthHelper, bool) {
+        let mut month_helper = MonthHelper {
+            day_number_month_starts_on: 0,
+            number_of_days_in_month: 0,
+            selected_year,
+            default_year_if_error,
+            month,
+        };
+
+        let chosen_date_is_valid_one = month_helper.calculate_day_number_month_starts_on();
+        let chosen_date_is_valid_two = month_helper.calculate_number_of_days_in_month();
+
+        (
+            month_helper,
+            chosen_date_is_valid_one && chosen_date_is_valid_two,
+        )
+    }
+
+    pub fn get_month_name(&self) -> String {
+        let month_name = match self.month {
+            1 => "january",
+            2 => "february",
+            3 => "march",
+            4 => "april",
+            5 => "may",
+            6 => "june",
+            7 => "july",
+            8 => "august",
+            9 => "september",
+            10 => "october",
+            11 => "november",
+            _ => "december",
+        };
+
+        month_name.to_string()
+    }
+
+    fn calculate_day_number_month_starts_on(&mut self) -> bool {
```

Comment:
A few of these I think read better as functions rather than methods, at least with the names you use.  This one is called `calculate`, but it returns a boolean, so I had to read the body a few times to see what was going on.

Would you be up using a hybrid of the approach you started with and the cleaner "all-in-one" structure of this PR?

I think you could keep the functions you had, and pass around the struct.

Alternatively, if you want to go with the methods, I think you could call them `update_*` and return a `Result` to say if the change was valid or not.

Reply:
Oh yeah, I forgot to rename those functions from when they actually returned a value and a boolean.  They should probably have the word `update` as you said.  As for the `Result` return value.. the boolean that returns has to do with whether the user input an invalid year or not.  If they input in an invalid year and chrono breaks, then the current year is used, so even if the year they put in fails, a valid year will be used.  The boolean returns just is used to tell the calling functions that they should also use the current year.

Are you suggesting just switching the bool for the result and doing the same exact logic?  Or doing something different?  I would've probably used a result originally, but a result only contains a valid if it is OK where this has a value in both cases of valid and invalid input.